src/myrtlespeech/model/hard_lstm.py

```diff
@@ -0,0 +1,558 @@
+        cn_out = torch.empty(0, device=self.device)
+        for rnn_layer in self.layers:
+            output, (h, c) = rnn_layer(output, (hn[i], cn[i]))
+            if i == 0:
+                hn_out = h.unsqueeze(0)
+                cn_out = c.unsqueeze(0)
+            else:
+                cn_out = torch.cat([cn_out, c.unsqueeze(0)], 0)
+                hn_out = torch.cat([hn_out, h.unsqueeze(0)], 0)
+            i += 1
+        req_size_out = (-1, batch, self._hidden_size)
+        out_states = hn_out.view(req_size_out), cn_out.view(req_size_out)
+
+        if self._batch_first:
+            output = output.transpose(0, 1)
+
+        return output, out_states
+
+
+class HardLSTMLayer(torch.nn.Module):
+    """Hard LSTM layer.
+
+    Args:
+        See :py:class:`HardLSTM`.
+    """
+
+    input_size: Final[int]
+    hidden_size: Final[int]
+    device: Final[torch.device]
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        forget_gate_bias: Optional[float] = None,
+    ):
+        super().__init__()
+        self.input_size = input_size
+        self.hidden_size = hidden_size
+        self.cell = HardLSTMCell(input_size, hidden_size, forget_gate_bias)
+
+        self.use_cuda = torch.cuda.is_available()
+
+        if self.use_cuda:
+            self.cell = self.cell.cuda()
+
+        self.device = next(self.cell.parameters()).device
+
+    def forward(
+        self, x: Tensor, hx: Tuple[Tensor, Tensor]
+    ) -> Tuple[Tensor, Tuple[Tensor, Tensor]]:
+        """Returns the result of applying the layer to ``(x, hx)``.
+
+        All inputs are moved to the GPU with :py:meth:`torch.nn.Module.cuda`
+        if :py:func:`torch.cuda.is_available` was :py:data:`True` on
+        initialisation.
+
+        Args:
+            x: A :py:class:`torch.Tensor` of rnn sequence inputs of size
+                ``[seq_len, batch, self.input_size]``.
+
+            hx: The lstm hidden state: a Tuple of lstm hidden state and cell
+                states where both are :py:class:`torch.Tensor`s of size
+                ``[1, batch, self.hidden_size]``.
+
+        Returns:
+            A Tuple of Tuples; ``res = a, (b, c)`` where all elements are
+            :py:class:`torch.Tensor`s. ``a`` is the lstm output of size
+            ``[seq_len, batch, self.hidden_size]`` and ``(c, d)`` are the
+            final lstm hidden and cell states.
+        """
+        hx = hx[0].squeeze(0), hx[1].squeeze(0)
+        timesteps = x.size(0)
+        # We extend y on each timestep with y = torch.cat([y, ...]) so
+        # must define initial y. Normally we could use y = torch.empty(0) but
+        # with onnx export, the dimensions must match for **all** concated
+        # tensors, even when empty. Hence define with correct shape but this
+        # means that it will not be treated as empty and will have to be
+        # removed from y later.
+        y = torch.empty(1, x.size(1), self.hidden_size, device=self.device)
```

Comment:
by using the `i==1` discussed in the previous comment. I've updated the one above so that they at least match (and moved the comment here to the `StackedLSTM` class).